Empty Stats: Add event tracking for Publicize / Blogging Reminders

momo-ozawa opened the pull request:
Part of #17307

## Description

This PR adds event tracking for the following events:
 
```
// Stats - Empty Stats nudges

stats_publicize_nudge_shown
stats_publicize_nudge_tapped
stats_publicize_nudge_dismissed
stats_publicize_nudge_completed
stats_blogging_reminders_nudge_shown
stats_blogging_reminders_nudge_tapped
stats_blogging_reminders_nudge_dismissed
stats_blogging_reminders_nudge_completed

// Stats - Customize card

stats_customize_insights_shown
```

## To test

To review this PR, please do the following:
1. Make sure events fire correctly by following the trigger in the table below
2. Validate the events (PCYsg-ktp-p2)

> What does it mean for an event to be valid?
> 
> An event can be marked valid when you’ve checked that it:
> 
>     Fires when it’s supposed to.
>     Fires exactly once each time it’s supposed to fire.
>     Has a blogid set if the event takes place in a site context.  If it’s not in a site context (e.g. a landing page or the wpcom homepage), you can omit a blogid.
>     Has the event properties that you intend it to have, and that are being attached to the event by your code.
> 
> Who should review it?
> 
> Similarly to code reviews, someone on your team or working in the same area as you is probably best to review your event.


## Event table

Event | Trigger | blog_id?
-- | -- | --
stats_publicize_nudge_shown | Publicize card is shown upon visiting the Stats Insights screen | ✅ 
stats_publicize_nudge_tapped  | Tapping on the `Enable post sharing` button on the Publicize card | ✅ 
stats_publicize_nudge_dismissed | Tapping on the `Dismiss` button on the Publicize card | ✅ 
stats_publicize_nudge_completed | Successfully setting up at least one connection via Publicize | ✅ 
stats_blogging_reminders_nudge_shown | Blogging Reminders card is shown upon visiting the Stats Insights screen | ✅ 
stats_blogging_reminders_nudge_tapped | Tapping on the `Set up blogging reminders` button on the Blogging Reminders card | ✅ 
stats_blogging_reminders_nudge_dismissed | Tapping on the `Dismiss` button on the Blogging Reminders card | ✅ 
stats_blogging_reminders_nudge_completed | Successfully setting up Blogging Reminders via the Blogging Reminders flow | ✅ 
stats_customize_insights_shown | Customize Insights card is shown upon visiting the Stats Insights screen | ❌ 

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested and made sure events fire properly with the correct triggers / properties

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add: events for publicize nudge, blogging reminders nudge, and customize card

Commit:
Add: extension method to track nudge event

Commit:
Add: track nudge tapped for publicize

Commit:
Add: track nudge tapped for blogging reminders

Commit:
Add: track nudge completed for publicize

Commit:
Add: track nudge completed for blogging reminders

momo-ozawa commented:
## Notes
- I'll add the `nudge_dismissed` events after Nikola's fix presentation logic PR lands. The plan is to track the events in the individual dismiss nudge methods for `SiteStatsInsightsDelegate` -> ✅

Commit:
Add: track card shown for customize insights

Commit:
Add: track nudge shown for publicize and blogging reminders

Commit:
Add: track nudge dismissed for publicize and blogging reminders

Commit:
Fix: return nil for itemToDisplay if view count is also nil

The default view count 0 was being supplied to itemToDisplay, which was causing .social to be returned when it should have been returning nil.

nikola-milicevic commented on WordPress/Classes/ViewRelated/Stats/Insights/SiteStatsInsightsTableViewController.swift:
```diff
@@ -211,18 +211,20 @@ private extension SiteStatsInsightsTableViewController {
         case .some(let item):
             switch item {
-            case is GrowAudienceCell.HintType where !insightsToShow.contains(.growAudience):
+            case let hintType as GrowAudienceCell.HintType where !insightsToShow.contains(.growAudience):
```
👌

Commit:
Revert "Fix: return nil for itemToDisplay if view count is also nil"

This reverts commit b4fb243fe419db5cf8b771d11b14d2799123b244.

Commit:
Fix: workaround to make sure shown events are fired only once

momo-ozawa commented:
Reverted b4fb243 because it was causing the presentation logic to break.
Added a workaround in c74d496 to make sure the `shown` events are fired only once.

nikola-milicevic reviewed (commented):
Great work!
Since you've already cleaned up some of my code 😄 could you just please get rid of nil coalescing operator [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/b4fb243fe419db5cf8b771d11b14d2799123b244/WordPress/Classes/ViewRelated/Stats/Insights/SiteStatsInsightsTableViewController.swift#L245)?
And `case let` syntax would be [great](https://github.com/wordpress-mobile/WordPress-iOS/blob/b4fb243fe419db5cf8b771d11b14d2799123b244/WordPress/Classes/ViewRelated/Stats/Insights/SiteStatsPinnedItemStore.swift#L58) here as well! (so I don't have to do [this](https://github.com/wordpress-mobile/WordPress-iOS/blob/b4fb243fe419db5cf8b771d11b14d2799123b244/WordPress/Classes/ViewRelated/Stats/Insights/SiteStatsPinnedItemStore.swift#L59))

I have just two concerns:
- `stats_blogging_reminders_nudge_shown` wasn't triggered for me 😬, I will check again (few times)
- `didChangePublicizeServices` method of `SharingViewControllerDelegate` could fire even if there are no changes (but that's not the scope of this PR - just pointing out here)

leandroalonso commented on WordPress/Classes/ViewRelated/Stats/Insights/SiteStatsInsightsTableViewController.swift:
```diff
@@ -211,18 +211,30 @@ private extension SiteStatsInsightsTableViewController {
                 insightsToShow = insightsToShow.filter { $0 != .growAudience }
                 insightsToShow.insert(.customize, at: 0)
+
+                // Work around to make sure customize insights shown is tracked only once
```
Is this to prevent the event from being fired multiple times?

if that's the case, this is something we should not be worried about, given that normally we look at "unique" events.

momo-ozawa replied:
Actually, this is a workaround to make sure the _correct_ event will be fired _once_.

If you go to a site with more than 30 views (i.e. customize should be showing) and comment out the `if viewsCount != nilstatement`, you'll see that the `publicize_nudge_shown` event is fired right before the correct event, `customize_insights_shown`. 

I tried doing b4fb243 in an attempt to fix this behavior, but that actually ended up breaking the presentation logic. (after dismissing a card, the next card shown would be the incorrect card, etc.) Maybe @nikola-milicevic has some insight on how to address this issue without breaking the presentation logic?

leandroalonso replied:
Oh ok! If that's the case then don't worry. Just wanted to make sure.

leandroalonso reviewed (changes requested):
@momo-ozawa it worked just fine for all the described events, except for the blogging reminders nudge appearing. No events are fired at all.

At least here, inside `loadPinnedCards` the execution is hitting the `break`. Let me know if you need more info!

nikola-milicevic reviewed (approved):
> stats_blogging_reminders_nudge_shown wasn't triggered for me 😬, I will check again (few times)

Checked again, few times, all good 😊
Not sure what happened.

leandroalonso commented:
> stats_blogging_reminders_nudge_shown wasn't triggered for me 😬, I will check again (few times)

I have the same issue 😬

leandroalonso commented:
FYI:

<img width="1512" alt="Screen Shot 2021-10-15 at 09 16 52" src="https://user-images.githubusercontent.com/7040243/137485338-a79bb0d6-f527-496c-af65-318076dd4fe9.png">

momo-ozawa commented:
> @momo-ozawa it worked just fine for all the described events, except for the blogging reminders nudge appearing. No events are fired at all.
> 
> At least here, inside `loadPinnedCards` the execution is hitting the `break`. Let me know if you need more info!

🤯 Hmmmm. Looking into this. So weird -- in your screenshot `item` evaluates to `.bloggingReminders`, which should be caught by the `case let hintType as GrowAudienceCell.HintType...` statement. Hmmmmm.

Commit:
Change RELEASE-NOTES

Commit:
Fix: remove growAudience from insightsToShow if nudge is completed

This fixes an issue where the following condition was failing:

where !insightsToShow.contains(.growAudience)

Commit:
Update: add relevant PRs to release notes

momo-ozawa commented:
@leandroalonso @nikola-milicevic 
Ready for a re-review! The mysterious bug where `blogging_reminders_nudge_shown` wasn't being fired has been fixed! 
94d4b9e

leandroalonso reviewed (approved):
@momo-ozawa :shipit: 

I also validated all the events.
